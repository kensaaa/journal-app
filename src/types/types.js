//lo que elta entre parentesis apunta al reducer que ejecuto la action
// [Auth] Login , apunta al reducer de authReducer


export const types = {
    login: '[Auth] Login',
    logout: '[Auth] Logout',

    uiSetError: '[UI] Set Error',
    uiRemoveError: '[UI] Remove Error',
    uiStartLoading: '[UI] Start Loading',
    uiFinishLoading: '[UI] Finish Loading',


    notesAddNew: '[Notes] New note',
    notesActive: '[Notes] Set active note',
    notesLoad: '[Notes] Load notes',
    notesUpdated: '[Notes] Update note',
    notesFileUrl: '[Notes] Update image url',
    notesDelete: '[Notes] Delete note',
    notesLogoutCleaning: '[Notes] Logout Cleaning',
}
