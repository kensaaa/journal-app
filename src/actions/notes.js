import Swal from "sweetalert2"
import {db} from "../firebase/firebase-config"
import {fileUpload} from "../helpers/fileUpload"
import {loadNotes} from "../helpers/loadNotes"
import {types} from "../types/types"


//es una tarea asincrona por eso el return de un callback
//
//cosas necesarias para grabar en firestore:
//
// uid del user ( para obtenerlo vamos usar el segundo argumento de callback getState, que obtiene el state )
//
//
export const startNewNote = () => {

    return async ( dispatch, getState ) => {

        const uid = getState().auth.uid

        const newNote = {
            title:'',
            body:'',
            date: new Date().getTime()
        }

        try{
            const doc = await db.collection(`${uid}/journal/notes`).add( newNote )

            //queremos que la  nota se active cuando se cree
            dispatch ( activeNote(doc.id, newNote) )
            dispatch ( addNewNote(doc.id, newNote) )
        }catch(err){
            console.log(err)
        }
        
    }

}

export const activeNote = (id, note) => ({
    type: types.notesActive,
    payload: {
        id,
        ...note
    }
})


export const addNewNote = (id,note) => ({
    type:types.notesAddNew,
    payload: {
        id,...note
    }
})


export const startLoadingNotes =  ( uid ) => {
    return async(dispatch) => {
        const notes = await loadNotes( uid )
        dispatch( setNotes( notes ) )
    }
}


export const setNotes = (notes) => ( {
    type: types.notesLoad,
    payload: notes 
})

//tarea asincrona , trabajamos con el midware thunk
export const startSaveNote = (note) => {
    return async(dispatch,getState) => {

        const { uid } = getState().auth

        if (!note.url){
            delete note.url
        }


        const noteToFirestore = { ...note }
        delete noteToFirestore.id

        await db.doc(`${uid}/journal/notes/${note.id}`).update(noteToFirestore)


        dispatch( refreshNote(note.id,noteToFirestore) )
        Swal.fire('saved',note.title,'success')
    }
}




//necesitamos el id de la nota y la nota, ( se pone el id por separado porque no siempre voy a tener el id de la nota )
//esta accion solo actualiza una nota
export const refreshNote = (id,note) => ({
    type: types.notesUpdated,
    payload: {
        id,
        //solucionar el problema del key
        note: {
            id,
            ...note
        }
    }
})



export const startUploading = (file) => {

    return async(dispatch,getState) => {

        //obtenemos la nota activa actual
        const { active:activeNote }  = getState().notes

        Swal.fire({
            title:'Uploading...',
            text:'Please wait...',
            //para sacar el botton de confirmacion
            showConfirmButton:false,
            //para no cancelar si hago click fuera del recuadro blanco
            allowOutsideClick:false,
            //esto ejecuta
            //un louding que no se pueda cerrar
            willOpen: () => {
                Swal.showLoading()
            }
        })

        const fileUrl = await fileUpload(file)
        activeNote.url = fileUrl

        //se guarda la imagen con la url 
        dispatch( startSaveNote( activeNote ) )


        //quitamos el para quitar el louding
        Swal.close()
    }
}


export const startDeleting = (id) => {

    return async(dispatch,getState) => {

        const uid = getState().auth.uid

        //lo borra de la base de datos
        await db.doc(`${uid}/journal/notes/${id}`).delete()
        //borrar fisicamente de nuestro listado y store
        dispatch( deleteNode(id) )

    }
}


export const deleteNode = (id) => ({
    type: types.notesDelete,
    payload: id
})


export const noteLogout = () => ({ type: types.notesLogoutCleaning })





