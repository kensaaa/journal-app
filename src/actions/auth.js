import Swal from 'sweetalert2'
import { firebase, googleAuthProvider } from '../firebase/firebase-config'
import {types} from "../types/types"
import {noteLogout} from './notes'
import {finishLoading, startLoading} from './ui'

export const startLoginEmailPassword = (email,password) => {

    return (dispatch) => {
        dispatch( startLoading() )
        //agremamos esto
        return firebase.auth().signInWithEmailAndPassword(email, password)
            .then( ({user}) => {

                dispatch( finishLoading() )
                dispatch( login(user.uid,user.displayName) )
            })
            .catch(e => {
                dispatch( finishLoading() )
                Swal.fire('Error',e.message,'error')
            })

    }
}

export const startGoogleLogin = () => {

    return (dispatch) => {

        //el signInWithPopup nesecita un provider que lo configuramos en firebase-config
        firebase.auth().signInWithPopup(googleAuthProvider)
            //cualdo se resuelve la promesa me entrega un userCred con varias cosas interesantes pero,
            // solo me importa la informcion del user que esta dentro del userCred
            .then( ({ user }) => {
                dispatch( login(user.uid, user.displayName) )
            })

    }

}

//es una tarea asincrona
export const registerWithEmailPasswordName = (email,password,name) => {

    return ( dispatch ) => {

        firebase.auth().createUserWithEmailAndPassword(email, password)
        //esto resueve un userCredential  lo mismo que con google
            .then( async ({ user }) => {

                //actualizamos al user, para colocar su nombre (tambien se puede la fotografia)
                //regresa una promesa( para no tener tantas promesas en cadena usamos async )
                await user.updateProfile({
                    displayName: name,
                })

                dispatch( login(user.uid, user.displayName) )
            })
        //existe error lo imprimo por consola
            .catch(e => {
                Swal.fire('Error', e.message, 'error')
            })

    }

}




export const login = (uid, displayName) => ({
    type: types.login,
    payload: {
        uid,
        displayName,
    }
})


//esteo es asincrono por eso callback
//esto action saca el usuario en firebase
export const startLogout = () => {
    return async (dispatch) => {
        //regresa una promesa
        await firebase.auth().signOut()

        dispatch( logout() )
        dispatch( noteLogout() )
    }
}


//esta action saca el usuario del store

export const logout = () => ({
    type: types.logout,
})
