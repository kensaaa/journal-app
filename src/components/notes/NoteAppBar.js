import React from 'react'
import {useDispatch, useSelector} from 'react-redux'
import {startSaveNote, startUploading} from '../../actions/notes'

const NoteAppBar = () => {

    const dispatch = useDispatch()
    const { active:note } = useSelector(state => state.notes)

    const handleSave = () => {
        dispatch( startSaveNote( note ) )	
    }

    const handlePictureClick = () => {
        //que busque el input y simule el click
        document.querySelector('#fileSelector').click()
    }

    //es igual que los inputs
    const handleFileChange = (e) => {
        //las imagenes se encuentran en e.target.files, ahi se encuentran todas las imagenes seleccionada
        //el user cancela el largo es de 0
        // console.log(e.target.files)
        
        const file = e.target.files[0]
        //si tenemos un file
        if (file){
            dispatch( startUploading( file ) )
        }
        
    }

    return (
        <div className='notes__appbar'>
            <span>28 agosto de 2020 </span>

            <input 
                id='fileSelector'
                type="file" 
                style={{
                    //para ocultarlo
                    display: 'none'
                }}
                //cuando cambia
                onChange={ handleFileChange }
            />

            <div>
                <button 
                    className="btn"
                    onClick={ handlePictureClick }
                >
                    Picture
                </button>

                <button 
                    className="btn"
                    onClick= { handleSave }
                >
                    Save
                </button>
            </div>
        </div>
    )
}

export default NoteAppBar
