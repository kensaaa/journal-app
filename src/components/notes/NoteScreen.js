import React from 'react'
import NoteAppBar from './NoteAppBar'

const NoteScreen = () => {

    return (
        <div className='notes__main-content'>

            <NoteAppBar />

            <div className="notes__content">

                <input 
                    type="text"
                    placeholder='Some awesome title'
                    className='notes__title-input'
                    autoComplete='off'
                />

                <textarea
                    placeholder='what happened today'
                    className='notes__text-area'
                    autoComplete='off'
                >
                </textarea>


                <div className="notes__image">
                    <img 
                        src="https://fondosmil.com/fondo/17010.jpg" 
                        alt="imagen" 
                    />

                </div>

            </div>


        </div>
    )
}

export default NoteScreen
