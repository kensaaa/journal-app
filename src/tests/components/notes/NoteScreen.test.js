import React from 'react'
import configureStore from 'redux-mock-store' //ES6 modules
import thunk from 'redux-thunk'
import { mount } from 'enzyme'
import {Provider} from 'react-redux'
import { MemoryRouter } from 'react-router-dom'

import {types} from '../../../types/types'
import NoteScreen from '../../../components/notes/NoteScreen'
import {activeNote} from '../../../actions/notes'

activeNote
jest.mock('../../../actions/notes',() => ({
    activeNote: jest.fn(),
}))


const middlewares = [thunk]
const mockStore = configureStore(middlewares)

const initState ={ 
    auth:{
        user:{
            uid:'ABC123',
            displayName:'juan'
        }
    },
    ui: {
        loading: false,
        msgError:null,
    },
    notes:{
        active:{
            id:1234,
            title:'hola',
            body:'mundo',
            date:0
        },
        notes:[]
    }
}

let store = mockStore( initState )
store.dispatch = jest.fn()

const wrapper = mount(
    <Provider store={store}>
        <MemoryRouter>
            <NoteScreen />
        </MemoryRouter>
    </Provider>

)

describe('Prueba en <NoteScreen />', () => {

    test('debe de mostrarse correctamente', () => {
        expect(wrapper).toMatchSnapshot()
    })


    test('debe de disparar el active note', () => {

        wrapper.find('input[name="title"]').simulate('change',{
            target:{
                name:'title',
                value:'hola de nuevo'
            }
        })

        //esto se ejecuta 2 veces porque activeNote esta dentro de un efecto, y la
        //primera vez no me importa por eso llamo al toHaveBeenLastCalledWith
        expect(activeNote).toHaveBeenLastCalledWith(
            1234,
            {
                body:'mundo',
                title: 'hola de nuevo',
                id:1234,
                date:0
            }
        )
        
    })
})
