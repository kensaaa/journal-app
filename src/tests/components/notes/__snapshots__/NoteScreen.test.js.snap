// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Prueba en <NoteScreen /> debe de mostrarse correctamente 1`] = `
<div
  className="notes__main-content"
>
  <div
    className="notes__appbar"
  >
    <span>
      28 agosto de 2020 
    </span>
    <input
      id="fileSelector"
      onChange={[Function]}
      style={
        Object {
          "display": "none",
        }
      }
      type="file"
    />
    <div>
      <button
        className="btn"
        onClick={[Function]}
      >
        Picture
      </button>
      <button
        className="btn"
        onClick={[Function]}
      >
        Save
      </button>
    </div>
  </div>
  <div
    className="notes__content"
  >
    <input
      autoComplete="off"
      className="notes__title-input"
      name="title"
      onChange={[Function]}
      placeholder="Some awesome title"
      type="text"
      value="hola"
    />
    <textarea
      autoComplete="off"
      className="notes__text-area"
      name="body"
      onChange={[Function]}
      placeholder="what happened today"
      value="mundo"
    />
  </div>
  <button
    className="btn btn-danger"
    onClick={[Function]}
  >
    delete
  </button>
</div>
`;
